### PR TITLE
Add ShipDesignManfest to parser.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR depends on #1588

It adds a parser for ShipDesignManifest which stores the order of ship designs in a ship design directory such as the default designs, monsters or saved designs.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
Could you rebase?

**Comment by LGM-Doyle** [comment]
rebased.

File: default/scripting/monster_designs/ShipDesignManifest.focs.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+ShipDesignManifest
```
Comment (geoffthemedio) [inline default/scripting/monster_designs/ShipDesignManifest.focs.txt]: Could you add a comment at the start of these files, indicating what they are for?

File: UI/ChatWnd.cpp [hunk 2]
```diff
@@ -195,13 +195,13 @@ void MessageWndEdit::FindGameWords() {
             m_game_words.insert(UserString(entry.second->Name()));
     }
     // add ship hulls
-    for (const auto& entry : GetPredefinedShipDesignManager()) {
-        if (entry.second->Hull() != "")
-            m_game_words.insert(UserString(entry.second->Hull()));
+    for (const auto& design : GetPredefinedShipDesignManager().GetOrderedShipDesigns()) {
```
Comment (geoffthemedio) [inline UI/ChatWnd.cpp:198]: I'm confused about this whole block of code... why is it only adding hulls / parts that are present in designs in the predefined manager... I'd change this to add all parts / hulls. Not really relevant to this PR, though...

File: parse/Parse.h [hunk 3]
```diff
@@ -48,12 +48,17 @@ namespace parse {
 
     FO_PARSE_API bool ship_hulls(std::map<std::string, HullType*>& hulls);
 
+    /** Parse all ship designs in directory \p path, store them in \p designs. If a file exists
+        called ShipDesignManifest.focs.txt, parse it and store the order in \p ordering. */
```
Comment (geoffthemedio) [inline parse/Parse.h]: Does `ShipDesignManifest.focs.txt` do anything besides determine the order in which the designs are displayed? If not, I think another name would be clearer about its purpose...

File: parse/ParseImpl.h [hunk 4]
```diff
@@ -126,9 +126,13 @@ namespace parse { namespace detail {
 
         TraceLogger() << "Parse: Elapsed time to parse " << path.string() << " = " << (timer.elapsed() * 1000.0);
 
-        std::ptrdiff_t distance = std::distance(first, parse::detail::s_end);
+        // s_end is global and static.  In is wrong when multiple files are concurrently or
```
Comment (geoffthemedio) [inline parse/ParseImpl.h]: In is -> It is
This change looks like it should be committed to master immediately and independently of this PR

File: parse/ShipDesignsParser.cpp [hunk 5]
```diff
@@ -90,15 +89,13 @@ namespace {
...
-            qi::_r6_type _r6;
             qi::eps_type eps;
 
             const parse::lexer& tok = parse::lexer::instance();
 
             design_prefix
                 =    tok.ShipDesign_
-                >    parse::detail::label(Name_token)
-                >    tok.string        [ _pass = is_unique_(_r6, ShipDesign_token, _1), _r1 = _1 ]
+                >    parse::detail::label(Name_token) > tok.string [ _r1 = _1 ]
```
Comment (geoffthemedio) [inline parse/ShipDesignsParser.cpp:98]: is this related to the subject of this PR?

File: UI/ChatWnd.cpp (hunk 2 above)
Comment (geoffthemedio) [reply UI/ChatWnd.cpp:198]: Even if there is a reason to add only seen hulls, it's still not enough to handicap the feature when players can see not-yet-available hulls in the pedia and might want to refer to them by name.

File: parse/Parse.h (hunk 3 above)
Comment (geoffthemedio) [reply parse/Parse.h]: `ShipDesignDesignScreenListOrdering.focs.txt` or `DesignScreenShipDesignListOrdering.focs.txt` or just `ShipDesignListOrdering.focs.txt`

File: UI/DesignWnd.cpp [hunk 6]
```diff
@@ -144,14 +144,13 @@ namespace {
             if (!exists(saved_designs_dir))
                 return;
 
-            std::map<std::string, std::unique_ptr<ShipDesign>> file_designs;
-            parse::ship_designs(saved_designs_dir, file_designs);
+            std::vector<std::unique_ptr<ShipDesign>> file_designs;
+            std::vector<boost::uuids::uuid> ordering;
+            parse::ship_designs(saved_designs_dir, file_designs, ordering);
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp:149]: In my tests, this doesn't find any saved designs. `file_designs` remains empty (in the debugger) after the parse, despite having two saved design files in `C:\Users\Geoff\AppData\Roaming\FreeOrion\shipdesigns`. The designs list remains empty.

**Review by LGM-Doyle — COMMENTED** [review]
The issues are addressed.

**Comment by LGM-Doyle** [comment]
rebased
